Add support for Global Data Consent Agreement

arslanashraf7 opened the pull request:
#### Pre-Flight checklist

- [x] Screenshots and design review for any changes that affect layout or styling
  - [x] Desktop screenshots

- [x] Migrations
  - [x] Migration is backwards-compatible with current production code

- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested

#### What are the relevant tickets?
#2178 

#### What's this PR do?
- Adds support for Global Consent Agreements

#### How should this be manually tested?
It might be a good idea to read the comments on issue #2178 and then:
1) Create a company (Optional)
2) Create coupon for a specific company using `ecommerce/admin/coupons/` and download the coupon codes
3) Create a consent agreement for a specific company and select the `All Courses` checkbox
4) Now apply a code from the csv downloaded in step 2
5) You should see a `Consent Agreement Message` at the bottom of the checkout page and upon clicking `Data Sharing Policy` You should see the `Content` added while creating the agreement
6) Now create another Consent Agreement for the same company and keep the `All Courses` checkbox unchecked and add a specific course to the list
7) Follow the steps for checkout and this time upon clicking `Data Sharing Policy` you should see course-based consent for the company.
8) The above scenario for Global consent should work for a course that is newly created and isn't added in any `Consent Agreement` here (`admin/ecommerce/dataconsentagreement`)


#### Where should the reviewer start?
(Optional)

#### Any background context you want to provide?
You can follow the associated ticket for more context

#### Screenshots (if appropriate)
**Validation message for Either making the agreement global or associate it with course**

<img width="1391" alt="Screenshot 2021-05-05 at 5 22 30 PM" src="https://user-images.githubusercontent.com/34372316/117141253-c202e980-adc7-11eb-82f0-31ce728a052c.png">

**Validation message for an existing Global agreement for a company since we expect only one global agreement per company**
<img width="974" alt="Screenshot 2021-05-05 at 5 24 49 PM" src="https://user-images.githubusercontent.com/34372316/117141271-c62f0700-adc7-11eb-8187-f92ffe8b0efd.png">

**Checkout Page with Consent checkbox**
<img width="1251" alt="Screenshot 2021-05-05 at 5 25 55 PM" src="https://user-images.githubusercontent.com/34372316/117141272-c6c79d80-adc7-11eb-92ef-f8dc42f6459e.png">

**Checkout page with Course based consent (Case when there is a course based consent available)**
<img width="992" alt="Screenshot 2021-05-05 at 5 26 08 PM" src="https://user-images.githubusercontent.com/34372316/117141281-c7f8ca80-adc7-11eb-9f75-29a4cf809cc6.png">

**Checkout page with Global Consent (Case when there is no course based consent available)**

<img width="1262" alt="Screenshot 2021-05-05 at 5 27 11 PM" src="https://user-images.githubusercontent.com/34372316/117141286-c8916100-adc7-11eb-870d-5458ffda40ca.png">

arslanashraf7 commented:
@briangrossman, Can you please take a look at the screenshots I've attached to the PR description for your review. I've also set up the CI for this PR [here](https://xpro-ci-pr-2201.herokuapp.com) in case you'd like to test it in a test environment.

Let me know if you think anything needs to be changed or if there is anything overlooked.

briangrossman commented:
@arslanashraf7 
I ran a test on the CI server and it seemed to work. 

I created a Data Consent for TestCo for Digital Learning 100...
![Screen Shot 2021-05-05 at 5 32 03 PM](https://user-images.githubusercontent.com/21342568/117212211-1017ed00-adc8-11eb-82c9-e687d7abd779.png)

 and another for All Courses
![Screen Shot 2021-05-05 at 5 32 14 PM](https://user-images.githubusercontent.com/21342568/117212222-1312dd80-adc8-11eb-8cb2-24b611df8aa5.png)


From the checkout, I saw the data consent specifically for Digital Learning 100 when I tried to check out with digital learning 100
![Screen Shot 2021-05-05 at 5 30 19 PM](https://user-images.githubusercontent.com/21342568/117212339-36d62380-adc8-11eb-80cd-ef879f7815ff.png)

And I saw the global data consent for Digital Learning 200
![Screen Shot 2021-05-05 at 5 30 34 PM](https://user-images.githubusercontent.com/21342568/117212384-46556c80-adc8-11eb-9c92-8c7c0927c4ab.png)

And digital learning 300
![Screen Shot 2021-05-05 at 5 30 47 PM](https://user-images.githubusercontent.com/21342568/117212412-4eada780-adc8-11eb-8643-34cf1555249a.png)

arslanashraf7 commented:
@briangrossman Thanks for running a test on this, I believe the similar global consent behavior will work for a newly created course as well.

noisecapella commented on ecommerce/api.py:
```diff
@@ -1067,12 +1067,21 @@ def get_or_create_data_consent_users(basket):
                 )
 
+                if not agreements:
+                    # Ideally, There should always be only one global consent agreement for a company at maximum
```
Can we log an error if there is more than one global consent agreement found?

arslanashraf7 replied:
Added.

arslanashraf7 commented:
@noisecapella As suggested, I've added the error log for if there are multiple global consent agreements for a company along with a test. Also, I've updated the admin part to handle one more case to stop creating multiple global consents from there.

This is ready for review again.

noisecapella commented:
Functionality looks good to me

arslanashraf7 commented on ecommerce/admin.py:
```diff
@@ -494,6 +497,39 @@ class DataConsentUserAdmin(TimestampedModelAdmin):
+        if (
+            is_global
+            and DataConsentAgreement.objects.filter(company=company, is_global=True)
+            .exclude(pk__in=[self.instance.pk or -1])
```
I added this because `self.instance.pk` in case of a new object is `None` and doing `.exclude(pk__in=[None])` would return 0 results. But the approach you mentioned without using the list filter works fine. Thanks for pointing this out.

I've updated the PR.

arslanashraf7 commented:
> Functionality looks good to me

@noisecapella Thanks for reviewing this. I've made a small update based on your comment.
